### PR TITLE
Added Ceilometer & MongoDB

### PR DESCRIPTION
**Pull request by ddaskal**
Added 3 Ceilometer roles: ceilometer-common/control/data
Ceilometer is currently trimmed down to minimal offering
Ceilometer pipeline.yml only allows cpu & cpu_utils meters
Added Ceilometer with tcp rule to haproxy

Added 1 MongoDB role with tasks split into main.yml & primary.yml
MongoDB is scalable to add additional controllers to replica set
MongoDB will deploy with one arbiter on compute1
MongoDB init script is idempotent to rerun ursula without failure
MongoDB only installs mongod daemon service and mongo-shell
Added MongoDB repo with key for future authentication
MongoDB currently has collectd template, but is untested

### CODE REVIEW

File: envs/example/defaults.yml [hunk 1]
```diff
@@ -21,6 +21,7 @@ secrets:
   horizon_secret_key:           asdf
   glance_sync:      ADQ64XUQLUWH75M634RVBLP55RKPGGOWG
   monitor_password: asdf
+  telemetry_password: asdf
```
Comment (paulczar) [inline envs/example/defaults.yml]: what is the difference between this and the ceilometer password ( secrets.service_password ) ?

File: roles/ceilometer-common/meta/main.yml [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+  - role: monitoring-common
+  - role: logging-config
+    service: ceilometer
+    logdata: "{{ ceilometer.logs }}"
+  - role: openstack-source
+    project_name: ceilometer
+    project_rev: "{{ ceilometer.source.rev }}"
+    virtualenv: "{{ ceilometer.virtualenv }}"
+    python_requirements:
+      - { name: pymongo }
```
Comment (paulczar) [inline roles/ceilometer-common/meta/main.yml]: should we pin the version of this ?

File: envs/example/defaults.yml (hunk 1 above)
Comment (ddaskal) [inline envs/example/defaults.yml]: telemetry_password is meant for rabbit, whereas service_password is for os & keystone authentication. we can use service_password if it is preferred to keep the same password. check ceilometer.conf under [publisher] group

File: roles/ceilometer-common/meta/main.yml (hunk 2 above)
Comment (ddaskal) [inline roles/ceilometer-common/meta/main.yml]: We thought better not to pin the pymongo version since we don't pin the mongodb version (it will grow within version 3.0.x)

Comment (nirajdp76) [inline roles/ceilometer-common/meta/main.yml]: probably better to pin both mongo server and client version

**Comment by zrs233** [comment]
@ddaskal Can you squash this into one commit?

File: envs/example/defaults.yml (hunk 1 above)
Comment (j2sol) [inline envs/example/defaults.yml]: If we keep them separate, it would make sense to name the variable in a way that indicates it's for rabbit.

File: roles/ceilometer-common/tasks/main.yml [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    - event_definitions.yaml
+    - event_pipeline.yaml
+    - pipeline.yaml
+
+- include: monitoring.yml tags=monitoring,common
+
+- include: logging.yml
+  tags:
+    - logrotate
+    - logging
```
Comment (j2sol) [inline roles/ceilometer-common/tasks/main.yml]: You use two different styles here for indicating tags on an include. Please pick one.

File: roles/ceilometer-common/templates/etc/ceilometer/api_paste.ini [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+# Remove authtoken from the pipeline if you don't want to use keystone authentication
+[pipeline:main]
+pipeline = authtoken api-server
+
+[app:api-server]
+paste.app_factory = ceilometer.api.app:app_factory
+
+[filter:authtoken]
+paste.filter_factory = keystonemiddleware.auth_token:filter_factory
+
```
Comment (j2sol) [inline roles/ceilometer-common/templates/etc/ceilometer/api_paste.ini]: Does ceilometer need keystone authtoken details for it to be able to auth?

File: roles/ceilometer-common/templates/etc/ceilometer/ceilometer.conf [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+# If mongodb_replica_set is set, MongoReplicaSetClient will be used instead of MongoClient
+mongodb_replica_set=mongoreplica
+
+[api]
+host = 0.0.0.0
+port = 8888
+
+[keystone_authtoken]
+identity_uri = {{ endpoints.identity_uri }}
+auth_uri = https://{{ endpoints.keystone }}:5001/v2.0
```
Comment (j2sol) [inline roles/ceilometer-common/templates/etc/ceilometer/ceilometer.conf]: We have an `endpoints.auth_uri` as well.

File: roles/ceilometer-common/templates/etc/ceilometer/ceilometer.conf [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+signing_dir = /var/cache/ceilometer/api
+
+[clients]
+ca_file = {{ ceilometer.cafile }}
+
+[publisher]
+telemetry_secret = {{ secrets.telemetry_password }}
+
+[service_credentials]
+os_auth_url = https://{{ endpoints.keystone }}:5001/v2.0
```
Comment (j2sol) [inline roles/ceilometer-common/templates/etc/ceilometer/ceilometer.conf]: This could be `endpoints.auth_uri` again.

File: roles/ceilometer-control/handlers/main.yml [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+- name: restart ceilometer services
+  service: name={{ item }} state=restarted must_exist=false
+  when: restart|default('True')
+  with_items:
+    - ceilometer-api
+    - ceilometer-collector
+    - ceilometer-agent-notification
+    - ceilometer-agent-central
+    - ceilometer-alarm-evaluator
+    - ceilometer-alarm-notifier
```
Comment (j2sol) [inline roles/ceilometer-control/handlers/main.yml]: Would it make more sense to put the handlers in the ceilometer-common role? That way all the handlers can live in one place but be called by anything that lists ceilometer-common as a requirement.

File: roles/ceilometer-control/tasks/monitoring.yml [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+  sensu_process_check: service=ceilometer-agent-central
+  notify: restart sensu-client
+
+- name: ceilometer-alarm-evaluator process check
+  sensu_process_check: service=ceilometer-alarm-evaluator
+  notify: restart sensu-client
+
+- name: ceilometer-alarm-notifier process check
+  sensu_process_check: service=ceilometer-alarm-notifier
+  notify: restart sensu-client
```
Comment (j2sol) [inline roles/ceilometer-control/tasks/monitoring.yml]: All of these could be done as a single task with a `with_items` list of processes to check.

File: roles/ceilometer-data/tasks/main.yml [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+  upstart_service: name=ceilometer-agent-compute
+                   user=ceilometer
+                   cmd=/usr/local/bin/ceilometer-agent-compute
+  notify:
+    - restart ceilometer services
+
+- name: pip install libvirt-python in venv on compute
+  pip: name=libvirt-python virtualenv="{{ ceilometer.virtualenv }}"
+  notify:
+    - restart ceilometer services
```
Comment (j2sol) [inline roles/ceilometer-data/tasks/main.yml]: Why not do this bit as part of the source install? Does it harm anything to have libvirt-python in the venv on all the hosts?

Maybe this bit was copied from another role, where we have to install this big into the /package/ virtualenv because some install time compilation is required. In those cases, they are conditioned with a `when` statement.

File: roles/horizon/tasks/main.yml [hunk 10]
```diff
@@ -92,7 +92,7 @@
   shell: "{{ item }}"
   with_items:
     - "{{ 'horizon'|ursula_package_path(openstack_package_version) }}/bin/django-admin.py collectstatic --noinput"
-#    - "{{ 'horizon'|ursula_package_path(openstack_package_version) }}/bin/django-admin.py compress"
+    - "{{ 'horizon'|ursula_package_path(openstack_package_version) }}/bin/django-admin.py compress"
```
Comment (j2sol) [inline roles/horizon/tasks/main.yml]: This change appears unrelated to the PR.

File: roles/mongodb/tasks/main.yml [hunk 11]
```diff
@@ -0,0 +1,38 @@
+---
+- name: install mongodb packages
+  apt: pkg={{ item }} state=installed
+  with_items:
+    - mongodb-org-server=3.0.6
+    - mongodb-org-shell=3.0.6
```
Comment (j2sol) [inline roles/mongodb/tasks/main.yml]: Perhaps make the versions of these and pymongo a variable in defaults. Saves hunting for where to change them when version bumping is required.

File: roles/mongodb/tasks/main.yml [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+
+- name: start mongod service
+  service: name=mongod state=started
+
+- include: primary.yml
+  when: mongodb_primary==true and mongo_installed.changed
+
+- name: add mongodb secondary servers to replica set
+  command: mongo --host {{ endpoints.mongodb }} --eval rs.add(\"{{ primary_ip ~ ":" ~ mongodb.port | string}}\")
+  when: mongodb_primary==false and inventory_hostname in groups['mongo_db'] and mongo_installed.changed
```
Comment (j2sol) [inline roles/mongodb/tasks/main.yml]: Using `mongo_installed` seems unsafe. If we bump mongo version, the apt task will register a change, which will cause this command to run on what could be an already configured replica set.

File: roles/mongodb/tasks/main.yml [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+
+- name: add mongodb arbiter to replica set
+  command: mongo --host {{ endpoints.mongodb }} --eval rs.addArb(\"{{ primary_ip ~ ":" ~ mongodb.port | string}}\")
+  when: mongodb_primary==false and inventory_hostname in groups['mongo_arbiter'] and mongo_installed.changed
```
Comment (j2sol) [inline roles/mongodb/tasks/main.yml]: Shouldn't the arbiter be added within the arbiter role?

File: roles/mongodb/tasks/primary.yml [hunk 14]
```diff
@@ -0,0 +1,17 @@
+---
+- name: create ceilometer user/db setup script
+  template:
+    src: etc/mongodb/create_ceilometer_db.sh
+    dest: /tmp/create_ceilometer_db.sh
+    owner: root
+    group: root
+    mode: 0700
+  when: mongodb_primary==true
```
Comment (j2sol) [inline roles/mongodb/tasks/primary.yml]: Why copy + execute when you could just use the `script` module?

Further, is there anything happening here that the `mongodb_user` module couldn't do?

File: site.yml [hunk 15]
```diff
@@ -78,6 +78,36 @@
...
+
+- name: install mongodb on secondaries
+  gather_facts: force
+  hosts: mongo_db:!mongo_db[0]
+  vars:
+    mongodb_primary: False
+  roles:
+    - role: mongodb
+      tags: ['mongodb', 'mongodb-secondary']
+      when: ceilometer.enabled|bool
```
Comment (j2sol) [inline site.yml]: Instead of doing two plays, you could do a single play, and turn your checks for `mongodb_primary` into checks for `inventory_hostname == mongo_db[0]`

File: roles/ceilometer-common/tasks/main.yml (hunk 3 above)
Comment (ddaskal) [inline roles/ceilometer-common/tasks/main.yml]: I changed it to be consistent, but I noticed all other services do the same as me where monitoring is one line, and logging is spread out. I added in 'when' conditionals like the others too

File: roles/ceilometer-common/templates/etc/ceilometer/ceilometer.conf (hunk 5 above)
Comment (ddaskal) [inline roles/ceilometer-common/templates/etc/ceilometer/ceilometer.conf]: fixed

File: roles/ceilometer-common/templates/etc/ceilometer/ceilometer.conf (hunk 6 above)
Comment (ddaskal) [inline roles/ceilometer-common/templates/etc/ceilometer/ceilometer.conf]: fixed

File: roles/ceilometer-common/tasks/main.yml (hunk 3 above)
Comment (j2sol) [inline roles/ceilometer-common/tasks/main.yml]: A fair point that other roles mix the style. Gotta start fixing it somewhere :)

File: envs/example/defaults.yml (hunk 1 above)
Comment (cjchang) [inline envs/example/defaults.yml]: The telemetry_password is used to config the "telemetry_secret" in the ceilometer.conf. When a meter in one ceilometer component is sent to other ceilometer component, this secret needs to be passed with the meter for other component to verify that the meter is coming from a valid source. Dean, please just call it telemetry_secret so it is clear and consistent for what it is. Password will be a little confusing.

File: roles/mongodb/templates/etc/mongodb/mongod.conf [hunk 16]
```diff
@@ -0,0 +1,82 @@
...
+# need to create and set permissions for this directory manually,
+# e.g., if the parent directory isn't mutable by the mongodb user.
+dbpath=/var/lib/mongodb
+
+#where to log
+logpath=/var/log/mongodb/mongod.log
+logappend=true
+
+pidfilepath = /var/run/mongodb.pid
+fork = true
```
Comment (cjchang) [inline roles/mongodb/templates/etc/mongodb/mongod.conf]: Currently, the "service mongod status/stop/restart" does not work. Please comment out the line "fork=true" to fix this issue.

File: roles/ceilometer-common/tasks/monitoring.yml [hunk 17]
```diff
@@ -0,0 +1,10 @@
+---
+- name: ceilometer-api check
```
Comment (cjchang) [inline roles/ceilometer-common/tasks/monitoring.yml]: As we have discussed, there are more to monitor. Please add them.

Comment (cjchang) [inline roles/ceilometer-common/tasks/monitoring.yml]: Just saw other processes monitored in ceilometer-controller. Then here are for common. We have 6 processes in controllers and 1 process in computes. I guess all the monitors should be in either ceilometer-controller or ceilometer-data and nothing should be in ceilometer-common.

File: roles/ceilometer-control/handlers/main.yml (hunk 7 above)
Comment (ddaskal) [inline roles/ceilometer-control/handlers/main.yml]: The reason I have these separated is because these 6 services are only on the controller nodes, while the compute only has one service. So merging them all into one handler on the ceilometer-common role would throw an error, right?

File: roles/ceilometer-data/tasks/main.yml (hunk 9 above)
Comment (ddaskal) [inline roles/ceilometer-data/tasks/main.yml]: I am fine with moving libvirt-python to install on all nodes, but Niraj originally advised me to do it this way to avoid cluttering the controller nodes with one more install.

File: envs/example/defaults.yml (hunk 1 above)
Comment (ddaskal) [inline envs/example/defaults.yml]: fixed

File: roles/ceilometer-common/tasks/main.yml (hunk 3 above)
Comment (ddaskal) [inline roles/ceilometer-common/tasks/main.yml]: fixed

File: roles/ceilometer-control/tasks/monitoring.yml (hunk 8 above)
Comment (ddaskal) [inline roles/ceilometer-control/tasks/monitoring.yml]: fixed

File: roles/mongodb/tasks/main.yml (hunk 11 above)
Comment (ddaskal) [inline roles/mongodb/tasks/main.yml]: fixed

File: roles/mongodb/tasks/main.yml (hunk 13 above)
Comment (ddaskal) [inline roles/mongodb/tasks/main.yml]: For mongodb, we only have one role with differentiated tasks for the different hosts.

File: roles/mongodb/templates/etc/mongodb/mongod.conf (hunk 16 above)
Comment (ddaskal) [inline roles/mongodb/templates/etc/mongodb/mongod.conf]: fixed

File: site.yml (hunk 15 above)
Comment (ddaskal) [inline site.yml]: We currently do it like this to ensure that the replica set and db/users are instantiated first, then the secondaries and arbiter are run afterwards. if it was all one play, then the tasks would run in parallel and error

Comment (j2sol) [inline site.yml]: Each task is ran in parallel, however there are controls to make sure the task only runs on one host. Can you pin point the tasks that would fail if each task was ran in parallel?

Comment (ddaskal) [inline site.yml]: mongodb/tasks/primary.yml (only run on controller1 currently) needs to run and complete before the next two tasks are called in monogdb/tasks/main.yml (specifically the "add mongodb secondary servers to replica set" task and "add mongodb arbiter to replica set").

File: roles/ceilometer-control/handlers/main.yml (hunk 7 above)
Comment (ddaskal) [inline roles/ceilometer-control/handlers/main.yml]: I just noticed that must_exist=false.... so that's how we avoid that issue. making the change now

File: site.yml (hunk 15 above)
Comment (j2sol) [inline site.yml]: You already have a conditional set on the inclusion of `primary.yml`. That conditional is applied to every task within, so even if running in parallel the tasks would only execute on the primary node. I don't see a reason to break this into two role applications.

File: roles/ceilometer-data/tasks/main.yml (hunk 9 above)
Comment (ddaskal) [inline roles/ceilometer-data/tasks/main.yml]: I made it to install libvirt-python as part of a source install on all three nodes, and added in a when conditional here to only do on a package install (until ceilometer gets merged into giftwrap)

File: roles/ceilometer-common/defaults/main.yml [hunk 18]
```diff
@@ -0,0 +1,58 @@
...
+      - ceilometer-collector
+      - ceilometer-agent-notification
+      - ceilometer-agent-central
+      - ceilometer-alarm-evaluator
+      - ceilometer-alarm-notifier
+      - ceilometer-agent-compute
+  purge_frequency: 604800
+  pymongo_version: '3.0.3'
+  virtualenv: /opt/openstack/current/ceilometer
+#  virtualenv: /opt/stack/ceilometer/.venv
```
Comment (j2sol) [inline roles/ceilometer-common/defaults/main.yml]: What is this comment here for?

File: roles/ceilometer-common/tasks/monitoring.yml (hunk 17 above)
Comment (j2sol) [inline roles/ceilometer-common/tasks/monitoring.yml]: This process monitor, for ceilometer, should go into the ceilometer-control role.

File: roles/ceilometer-common/tasks/monitoring.yml [hunk 19]
```diff
@@ -0,0 +1,10 @@
...
+  sensu_check: name=check-ceilomter-api plugin=check-os-api.rb
```
Comment (j2sol) [inline roles/ceilometer-common/tasks/monitoring.yml]: ceilometer is spelled wrong on this line.

File: roles/mongodb/tasks/main.yml [hunk 20]
```diff
@@ -0,0 +1,42 @@
...
+  notify:
+    - restart mongod service
+
+- meta: flush_handlers
+
+- name: start mongod service
+  service: name=mongod state=started
+
+- include: primary.yml
+#  run_once: true
```
Comment (j2sol) [inline roles/mongodb/tasks/main.yml]: Remove comment.

File: roles/mongodb/tasks/main.yml [hunk 21]
```diff
@@ -0,0 +1,42 @@
...
+  when: inventory_hostname in groups['mongo_db'][0] #and mongo_installed.changed
```
Comment (j2sol) [inline roles/mongodb/tasks/main.yml]: Why is there commented out bits here?

File: roles/mongodb/tasks/main.yml [hunk 22]
```diff
@@ -0,0 +1,42 @@
...
+
+- name: add mongodb secondary servers to replica set
+  command: mongo --host {{ endpoints.mongodb }} --eval rs.add(\"{{ primary_ip ~ ":" ~ mongodb.port | string}}\")
+  when: inventory_hostname not in groups['mongo_db'][0] and inventory_hostname in groups['mongo_db'] #and mongo_installed.changed
+
+- name: add mongodb arbiter to replica set
+  command: mongo --host {{ endpoints.mongodb }} --eval rs.addArb(\"{{ primary_ip ~ ":" ~ mongodb.port | string}}\")
+  when: inventory_hostname in groups['mongo_arbiter'] #and mongo_installed.changed
```
Comment (j2sol) [inline roles/mongodb/tasks/main.yml]: Same question on these two lines.

File: roles/mongodb/tasks/primary.yml [hunk 23]
```diff
@@ -0,0 +1,35 @@
...
+#     group: root
+#     mode: 0700
+#   when: mongodb_primary==true
+#
+# - name: run ceilometer user/db setup script
+#   command: /tmp/create_ceilometer_db.sh
+#   register: create_result
+#   changed_when: "create_result.rc == 0"
+#   failed_when: "create_result.rc > 1"
+#   when: mongodb_primary==true
```
Comment (j2sol) [inline roles/mongodb/tasks/primary.yml]: Why is this all commented out?

**Comment by j2sol** [comment]
retest

File: roles/ceilometer-data/tasks/main.yml [hunk 24]
```diff
@@ -0,0 +1,24 @@
...
+                   user=ceilometer
+                   cmd=/usr/local/bin/ceilometer-agent-compute
+  notify:
+    - restart ceilometer services
+
+- name: pip install libvirt-python in venv on compute
+  pip: name=libvirt-python virtualenv={{ ceilometer.virtualenv }}
+  notify:
+    - restart ceilometer services
+  when: openstack_install_method == 'package'
```
Comment (j2sol) [inline roles/ceilometer-data/tasks/main.yml]: Don't we need libvirt-python in the virtualenv on source installs too?

File: roles/mongodb/tasks/main.yml [hunk 25]
```diff
@@ -0,0 +1,40 @@
+---
+- name: install mongodb packages
+  apt: pkg={{ item }} state=installed
+  with_items:
+    - mongodb-org-server={{ mongodb.version }}
+    - mongodb-org-shell={{ mongodb.version }}
```
Comment (j2sol) [inline roles/mongodb/tasks/main.yml]: Will the package upgrade cause a service restart, or should this task also trigger a restart of the service (when a mongo version upgrade happens)?

File: site.yml [hunk 26]
```diff
@@ -78,6 +78,22 @@
     - role: percona-backup
       tags: ['infra', 'percona']
 
+- name: install mongodb on primary
```
Comment (j2sol) [inline site.yml]: This play name should change to reflect that it's for all hosts in the mongo_db group.

File: envs/example/standard/hosts [hunk 27]
```diff
@@ -10,9 +10,16 @@ controller2
...
 controller2
 
+[mongo_db]
+controller1
+controller2
+
 [db_arbiter]
 compute1
 
+[mongo_arbiter]
```
Comment (j2sol) [inline envs/example/standard/hosts]: Would it make sense to re-use the generic `db` and `db_arbiter` host groups for mongodb? We're going by the same rules, the db active/active on the controllers and passive on a compute.

**Comment by j2sol** [comment]
re-reading how we handle mysql, I think you should model the mongo roles similar.

A set of roles for installing mongo itself, modeled after percona common, server, arbiter roles. Then a role for creating the database within mongo for use by ceilometer, modeled after the openstack-database role, and called in a similar way. This keeps things consistent.

File: roles/ceilometer-data/tasks/main.yml (hunk 24 above)
Comment (ddaskal) [inline roles/ceilometer-data/tasks/main.yml]: When installing from source, libvirt-python is installed as a system dependency on the compute node as part of the nova-data role. However, for package install the nova-data role has to install it manually to the nova virtualenv (I mention link & lines below). So for package installs, I also had to install libvirt in my ceilometer virtualenv, since it wasn't installed system-wide.

https://github.com/blueboxgroup/ursula/blob/1c2e57c9b5382813b1f005fa28da814a54f222fd/roles/nova-data/tasks/libvirt.yml (lines 37-48)

Comment (nirajdp76) [inline roles/ceilometer-data/tasks/main.yml]: When installing from source nova, neutron, cinder and other projects (except for horizon) does not install in a virtualenv. We were following nova-common/meta/main.yml

File: envs/example/standard/hosts (hunk 27 above)
Comment (ddaskal) [inline envs/example/standard/hosts]: Initially that's how I had it, but Paul suggested that I separate it out so MongoDB is not tightly coupled with MySQL considering it is currently an optional service with Ceilometer. This also allows for future architectural flexibility as we scale. Let me know which way is better for now.

File: site.yml (hunk 26 above)
Comment (ddaskal) [inline site.yml]: done

File: roles/mongodb/tasks/main.yml (hunk 25 above)
Comment (ddaskal) [inline roles/mongodb/tasks/main.yml]: I just verified that manually upgrading mongo versions DOES trigger the mongod service to restart when upgrade is completed.

Therefore, I'm assuming if someone were to update the {{ mongodb.version }} after an initial install and rerun the deployment, then Ansible is smart enough to detect the version change and run that task, thus triggering mongodb to restart itself when complete.

File: envs/example/standard/hosts (hunk 27 above)
Comment (cjchang) [inline envs/example/standard/hosts]: It is more flexible to separate them. The db is used for MySQL for other components. Today, it is clear that we have 2 controllers and 1 compute with primary, secondary, and arbiter. When this configuration is changed, potentially the requirements for MySQL and Mongo may or may not be the same. With they are separated, we have one thing less to investigate and worry in the future.

File: roles/ceilometer-common/tasks/monitoring.yml [hunk 28]
```diff
@@ -0,0 +1,6 @@
+---
+- name: ceilometer sla metrics
```
Comment (cjchang) [inline roles/ceilometer-common/tasks/monitoring.yml]: This needs to go to the controller too.

Comment (cjchang) [inline roles/ceilometer-common/tasks/monitoring.yml]: there is no need to monitoring.yml for common...

File: roles/ceilometer-common/handlers/main.yml [hunk 29]
```diff
@@ -0,0 +1,13 @@
...
+  service: name={{ item }} state=restarted must_exist=false
+  when: restart|default('True')
+  with_items:
+    - ceilometer-api
+    - ceilometer-collector
+    - ceilometer-agent-notification
+    - ceilometer-agent-central
+    - ceilometer-alarm-evaluator
+    - ceilometer-alarm-notifier
+    - ceiloemter-agent-compute
```
Comment (cjchang) [inline roles/ceilometer-common/handlers/main.yml]: It is fine to put the handlers in the common role but we will need to separate it into two restart service. We will need one with "restart ceilometer controller service", which will restart only the 6 controller processes and one with "restart ceilometer data service", which will restart the 1 data process. In each controller and data main.yml, it will call its own corresponding restart.  Right now, it will try to restart all of them in controllers and compute. I believe the all codes are in all 3 nodes, it will end up all 7 processes are running on all 3 nodes, I think.

File: envs/example/defaults.yml [hunk 30]
```diff
@@ -21,6 +21,7 @@ secrets:
   horizon_secret_key:           asdf
   glance_sync:      ADQ64XUQLUWH75M634RVBLP55RKPGGOWG
   monitor_password: asdf
+  telemetry_secret: asdf
```
Comment (cjchang) [inline envs/example/defaults.yml]: A default value for "mongodb_password" will need to be added in the default.yml too. It is in all.yml only. the deployment should work without the value in all.yml so it is either it is in default.yml or the code can handle missing value.

File: envs/example/standard/hosts (hunk 27 above)
Comment (j2sol) [inline envs/example/standard/hosts]: I suppose I can see that.

File: envs/example/defaults.yml (hunk 30 above)
Comment (ddaskal) [inline envs/example/defaults.yml]: done

File: roles/ceilometer-common/tasks/main.yml (hunk 3 above)
Comment (ddaskal) [inline roles/ceilometer-common/tasks/main.yml]: done

File: roles/ceilometer-data/tasks/main.yml (hunk 24 above)
Comment (j2sol) [inline roles/ceilometer-data/tasks/main.yml]: This recently changed with a PR from Paul, and now all the services get their own virtualenv even when installing via source.

File: roles/ceilometer-common/handlers/main.yml (hunk 29 above)
Comment (craigtracey) [inline roles/ceilometer-common/handlers/main.yml]: I do not believe that this is true as must_exist=false.

File: roles/ceilometer-common/templates/etc/ceilometer/event_definitions.yaml [hunk 31]
```diff
@@ -0,0 +1,62 @@
+---
```
Comment (craigtracey) [inline roles/ceilometer-common/templates/etc/ceilometer/event_definitions.yaml]: If this is not a template it should be under "files"

File: roles/mongodb-server/tasks/primary.yml [hunk 32]
```diff
@@ -0,0 +1,18 @@
...
+- name: rename primary replica member
+  command: mongo --host {{ endpoints.mongodb }} --eval 'cfg=rs.conf(); cfg.members[0].host="{{ primary_ip }}:{{ mongodb.port }}"; rs.reconfig(cfg)'
+
+- name: install pymongo for mongodb_user module
+  pip: name=pymongo state=present
+
+- name: create admin user
+  mongodb_user: database=admin name=admin password={{ mongodb.ceilometer_password }} roles='userAdminAnyDatabase' state=present
+
+- name: create ceilometer user
```
Comment (craigtracey) [inline roles/mongodb-server/tasks/primary.yml]: Shouldn't this be done by ceilometer roles?

File: site.yml [hunk 33]
```diff
@@ -78,6 +78,22 @@
     - role: percona-backup
       tags: ['infra', 'percona']
 
+- name: install mongodb server
+  gather_facts: force
```
Comment (craigtracey) [inline site.yml]: this does not look to be necessary

File: site.yml [hunk 34]
```diff
@@ -78,6 +78,22 @@
...
+  hosts: mongo_db
+  roles:
+    - role: mongodb-server
+      tags: ['mongodb', 'mongodb-server']
+      when: ceilometer.enabled|bool
+
+- name: install mongodb arbiter
+  gather_facts: force
```
Comment (craigtracey) [inline site.yml]: also not necessary afaict

File: roles/ceilometer-common/handlers/main.yml (hunk 29 above)
Comment (ddaskal) [inline roles/ceilometer-common/handlers/main.yml]: correct - already tested successful

File: roles/ceilometer-common/templates/etc/ceilometer/event_definitions.yaml (hunk 31 above)
Comment (ddaskal) [inline roles/ceilometer-common/templates/etc/ceilometer/event_definitions.yaml]: I believe the reasoning for having it here was to consolidate all ceilometer input templates in one place, which then makes it easy to copy over to the nodes in a single task

File: roles/mongodb-server/tasks/primary.yml (hunk 32 above)
Comment (ddaskal) [inline roles/mongodb-server/tasks/primary.yml]: these commands are all mongo commands for setting up the mongodb replica set, and the mongo/ceilometer users/databases. I don't think executing a mongo command should be done from a different role

File: site.yml (hunk 34 above)
Comment (ddaskal) [inline site.yml]: the compute node isn't a server, just an arbiter. so we needed to split up into two plays for the two different host groups - similar to percona

File: site.yml (hunk 33 above)
Comment (ddaskal) [inline site.yml]: see below comment

File: roles/mongodb-server/tasks/primary.yml (hunk 32 above)
Comment (craigtracey) [inline roles/mongodb-server/tasks/primary.yml]: It should be done from ceilometer as ceilometer depends on mongo being present.  If, god forbid, there were another place we wanted to use mongo, we would not necessarily want the ceilometer user present.

File: site.yml (hunk 34 above)
Comment (craigtracey) [inline site.yml]: This is specific to "gather_facts: force"

File: site.yml (hunk 33 above)
Comment (ddaskal) [inline site.yml]: resolved

File: site.yml (hunk 34 above)
Comment (ddaskal) [inline site.yml]: resolved

**Comment by craigtracey** [comment]
retest

**Comment by ddaskal** [comment]
retest

**Comment by ddaskal** [comment]
retest

**Comment by ddaskal** [comment]
retest

**Comment by ddaskal** [comment]
retest

**Comment by ddaskal** [comment]
retest

**Comment by ddaskal** [comment]
retest

**Comment by ddaskal** [comment]
retest

**Comment by ddaskal** [comment]
retest

**Comment by ddaskal** [comment]
retest

**Comment by j2sol** [comment]
retest

**Comment by ddaskal** [comment]
retest

**Comment by j2sol** [comment]
This is going to fail until #1252 is merged and this is rebased on it.
